feat: Advanced file nesting with patterns

Teddytrombone opened the pull request:
Solves #349 
File nesting now can be configured to use a pattern for file matching instead of just relying on extensions. The old approach is still supported so this shouldn't be a breaking change.
The structure of such a rule is:
```lua
"key" = {
        pattern = "...",
        ignore_case = true -- optional
        files = { ... },
}
```
Each pattern can contain mathcing groups which then can be used for the nested file name matching.

Here's an advanced configuration example, which was successfully tested:

```lua
nesting_rules = {
    ["js"] = {
        pattern = "(.+)%.js$",
        files = { "%1.js.map", "%1.min.js", "%1.d.ts" },
    },
    [".gitignore"] = {
        pattern = "^%.gitignore$",
        files = { ".gitattributes", ".gitmodules", ".gitmessage", ".mailmap", ".git-blame*" },
    },
    ["package.json"] = {
        pattern = "^package%.json$",
        files = {
            "package-lock.json",
            "yarn*",
            ".browserslist*",
            ".circleci*",
            ".commitlint*",
            ".cz-config.js",
            ".czrc",
            ".dlint.json",
            ".dprint.json*",
            ".editorconfig",
            ".eslint*",
            ".firebase*",
            ".flowconfig",
            ".github*",
            ".gitlab*",
            ".gitpod*",
            ".huskyrc*",
            ".jslint*",
            ".lintstagedrc*",
            ".markdownlint*",
            ".node-version",
            ".nodemon*",
            ".npm*",
            ".nvmrc",
            ".pm2*",
            ".pnp.*",
            ".pnpm*",
            ".prettier*",
            ".release-please*.json",
            ".releaserc*",
            ".sentry*",
            ".simple-git-hooks*",
            ".stackblitz*",
            ".styleci*",
            ".stylelint*",
            ".tazerc*",
            ".textlint*",
            ".tool-versions",
            ".travis*",
            ".versionrc*",
            ".vscode*",
            ".watchman*",
            ".xo-config*",
            ".yamllint*",
            ".yarnrc*",
            "Procfile",
            "apollo.config.*",
            "appveyor*",
            "azure-pipelines*",
            "bower.json",
            "build.config.*",
            "bun.lockb",
            "commitlint*",
            "crowdin*",
            "dangerfile*",
            "dlint.json",
            "dprint.json*",
            "electron-builder.*",
            "eslint*",
            "firebase.json",
            "grunt*",
            "gulp*",
            "jenkins*",
            "lerna*",
            "lint-staged*",
            "nest-cli.*",
            "netlify*",
            "nodemon*",
            "npm-shrinkwrap.json",
            "nx.*",
            "package.nls*.json",
            "phpcs.xml",
            "pm2.*",
            "pnpm*",
            "prettier*",
            "pullapprove*",
            "pyrightconfig.json",
            "release-please*.json",
            "release-tasks.sh",
            "release.config.*",
            "renovate*",
            "rollup.config.*",
            "rspack*",
            "simple-git-hooks*",
            "stylelint*",
            "tslint*",
            "tsup.config.*",
            "turbo*",
            "typedoc*",
            "unlighthouse*",
            "vercel*",
            "vetur.config.*",
            "webpack*",
            "workspace.json",
            "xo.config.*",
        },
    },
    ["readme"] = {
        pattern = "^readme.*",
        ignore_case = true,
        files = {
            "authors",
            "backers*",
            "changelog*",
            "citation*",
            "code_of_conduct*",
            "codeowners",
            "contributing*",
            "contributors",
            "copying*",
            "credits",
            "governance.md",
            "history.md",
            "license*",
            "maintainers",
            "readme*",
            "security.md",
            "sponsors*",
        },
    },
    ["go"] = {
        pattern = "(.*)%.go$",
        files = { "%1_test.go" },
    },
    ["docker"] = {
        pattern = "^dockerfile$",
        ignore_case = true,
        files = { ".dockerignore", "docker-compose.*", "dockerfile*" },
    },
    ["composer.json"] = {
        pattern = "^composer%.json$",
        files = { ".php*.cache", "composer.lock", "phpunit.xml*", "psalm*.xml" },
    },
    [".env"] = {
        pattern = "^%.env$",
        files = { "*.env", ".env.*", ".envrc", "env.d.ts" },
    },
    ["vite.config"] = {
        pattern = "^vite%.config%..*",
        files = {
            "*.env",
            ".babelrc*",
            ".codecov",
            ".cssnanorc*",
            ".env.*",
            ".envrc",
            ".htmlnanorc*",
            ".lighthouserc.*",
            ".mocha*",
            ".postcssrc*",
            ".terserrc*",
            "api-extractor.json",
            "ava.config.*",
            "babel.config.*",
            "contentlayer.config.*",
            "cssnano.config.*",
            "cypress.*",
            "env.d.ts",
            "formkit.config.*",
            "formulate.config.*",
            "histoire.config.*",
            "htmlnanorc.*",
            "i18n.config.*",
            "jasmine.*",
            "jest.config.*",
            "jsconfig.*",
            "karma*",
            "lighthouserc.*",
            "playwright.config.*",
            "postcss.config.*",
            "puppeteer.config.*",
            "rspack.config.*",
            "svgo.config.*",
            "tailwind.config.*",
            "tsconfig.*",
            "tsdoc.*",
            "uno.config.*",
            "unocss.config.*",
            "vitest.config.*",
            "vuetify.config.*",
            "webpack.config.*",
            "windi.config.*",
        },
    },
    ["tex"] = {
        pattern = "^(.*)%.tex$",
        files = {
            "%1.acn",
            "%1.acr",
            "%1.alg",
            "%1.aux",
            "%1.bbl",
            "%1.blg",
            "%1.fdb_latexmk",
            "%1.fls",
            "%1.glg",
            "%1.glo",
            "%1.gls",
            "%1.idx",
            "%1.ind",
            "%1.ist",
            "%1.lof",
            "%1.log",
            "%1.lot",
            "%1.out",
            "%1.pdf",
            "%1.synctex.gz",
            "%1.toc",
            "%1.xdv",
        },
    },
    ["jsx"] = {
        pattern = "^(.*)%.jsx$",
        files = { "%1.js", "%1.*.jsx", "%1_*.js", "%1_*.jsx", "%1.less", "%1.module.less" },
    },
    ["ts"] = {
        pattern = "^(.*)%.ts",
        files = { "%1.js", "%1.d.ts.map", "%1.*.ts", "%1_*.js", "%1_*.ts" },
    },
    ["svg"] = { "svgz", "svg.br" }, -- configuration via extensions ist still possible
    ["css"] = { "css.map", "min.css" }, -- configuration via extensions ist still possible
}
``` 

At the moment there is one drawback: if there are colliding rules like "vite.config" and ".env" from the example above then it may be that multiple nested items occur. I think i depends on the order of the rules. Maybe some "flattening" has to happen after nesting but I have no clue, if this would be a good solution.

The above configuration leads to something like this:

Collapsed:
![example-collapsed](https://github.com/nvim-neo-tree/neo-tree.nvim/assets/14216933/cca5d896-d17c-40af-95cb-fdda1659acb9)

Uncollapsed
![example-uncollapsed](https://github.com/nvim-neo-tree/neo-tree.nvim/assets/14216933/15269f99-78b7-4dce-8d6b-78978fb6d55a)

cseickel reviewed (commented):
Thanks a lot for contributing and putting together such a well made PR with lots of supporting documentation.

There is definitely more work to do before I can accept it though. This is being run at a critical place and we can't accept any slowdowns here. We also can't allow this feature to add bugs or performance penalties for people that are not using it.

Teddytrombone commented on doc/neo-tree.txt:
```diff
@@ -1208,6 +1208,52 @@ This will render:
     FILENAME.js
         FILENAME.js.map
 
+A more advanced usage are rules with patterns. Those can be mixed with file
```
I'm not shure if I understand what exactly you mean. The patterns are regex-like Lua patterns (https://www.lua.org/pil/20.2.html) which can have captures (https://www.lua.org/pil/20.3.html). Via the captures we can check for parts of an filename to be used for nested files. e.g. in go, we can "capture" the filename and then determine, what file ist the test file.
If this is not what you meant please let me know.

cseickel replied:
I guess my confusion is that I thought these were global patterns and not lua patterns, because you are using globtopattern to convert them. Now I see that these are actually lua patterns so the `%` I was confused about were actually escaping the `.` to be a literal dot.

If that is the case, why are you using globtopattern?

We normally use that library because globs are pretty universal and lua patterns are not, so most people would prefer to use globs rather than learn a new syntax.

Teddytrombone replied:
I wanted the pattern matching to be more or less universal but also as easy to use as possible. So for file matching of the "main" file I use patterns so I can have captures. Those captures can then be used for matching the nested files, but thos have to be globs in order to have wildcards for different matching rules. As globs are easier to use if there are many files to match and it's easier to use the captures without breaking patterns. Maybe this is a bit overenginieered or inconsistent but it works ;-)
I will try to explain the "flow" with an example: 

```lua
"test" = {
    pattern = "foo_(.*)%.bar"
    files = { "foobar.*", "%1.sub", "child_%1.*" }
}
```

The pattern would match a file `foo_test.bar` with a capture `test`. After we have the capture we can insert it into the file globs: `foobar.*`, `test.sub`, `child_test.*`. These globs are then converted to patterns to match files like `foobar.md`, `test.sub`, `child_test.js`, `child_test.tmp`

I already have a TODO on my list to think about checking the nested file pattern if this is obviously a glob or just a fixed string. If the latter we could skip the globtopattern part and just compare the file name. But first I have to refactor my code ;-)

If you think this is too clunky let me know if you would prefer a different approach - or if you dont't like it at all just reject the pull request - no hard feelings 😊

cseickel replied:
OK, I get it. It is a definitely confusing that we are using two different types of patterns within the same table. I do understand why now. I don't have any better ideas, but I would say that if this is the only way then it needs to be spelled out very clearly that one is lua pattern and one is a glob pattern. I would include the explanation in the docs and also add a reminder anywhere an example is displayed. Like this:

``` lua
"test" = {
    pattern = "foo_(.*)%.bar", -- <-- Lua pattern
    files = { "foobar.*", "%1.sub", "child_%1.*" } -- <-- glob pattern
}
```

Teddytrombone commented on lua/neo-tree/sources/common/file-nesting.lua:
```diff
@@ -1,21 +1,75 @@
+            local item_name = item.name
+            if ruleConfig["ignore_case"] then
+              pattern = case_insensitive_pattern(pattern)
+              item_name = item_name:lower()
```
As far as i know there is no way in Lua to make a pattern cas insensitive like it is in POSIX regexes via modifiers. So either the user has to secify a case insensitive pattern via declaring each character as lower and uppercase like `[dD][oO][cC][kK][eE][rR][fF][iI][lL][eE]` or we have to handle this like I did it in the code.
Maybe there is a neovim function, which would handle this. If that's the case I would be glad if you could point me in the right direction ;-)

cseickel replied:
Ugh, you're right. That sucks. I guess `:lower()` is the way to go.

Teddytrombone commented on lua/neo-tree/sources/common/file-nesting.lua:
```diff
@@ -1,21 +1,75 @@
+  if item.type ~= "file" then
+    return nil
+  end
+  local dir = vim.loop.fs_scandir(item.parent_path)
```
Yes you are right. I already tried to accomplish this, but I got errors with the git portion of neo-tree. I will do my best to dig deeper into the whole plugin and see, if i find a solution.

cseickel replied:
If it were easy, it would have been done already. 😄

Teddytrombone commented:
Thank you for your kind and also detailed response. You made some good points and will review my code based on your suggestions. I think it will take some time until I got deeper into the original code base. But as soon as I am (hopefully) confident with it, I will try to make a better PR 🤔

Teddytrombone commented:
I completely refactored my code and also the existing nesting code. The code is now run on fs_scan -> job_complete as there are all files known. Please have a look at the code and let me know what you think.

cseickel commented:
There's a lot to review here but it definitely looks like you did a great job from a quick skim. It might take me a few days to get to it.

cseickel commented on lua/neo-tree/sources/common/file-nesting.lua:
```diff
@@ -25,30 +57,224 @@ function M.get_parent(item)
+  for _, matcher in pairs(matchers) do
+    if matcher.enabled then
+      if matcher.can_have_nesting(item) then
+        return
```
Should this be `return true`?

Teddytrombone replied:
You are right, this is legacy code which was replaced by get_nesting_callback. I will remove it.

cseickel reviewed (commented):
This is much improved, thank you. I think we are good to go except for the `can_have_nesting` function.

I have not actually tested it to see that it works as advertised, can you just confirm for me that you have?

- [ ] The original file nesting config still works the same
- [ ] The new config does everything the docs says it should
- [ ] Invalid patterns do not break Neo-tree 
- [ ] People that do not use file nesting will not be impacted at all

Teddytrombone commented:
- [x]  The original file nesting config still works the same
- [x]  The new config does everything the docs says it should
- [ ]  Invalid patterns do not break Neo-tree
That's a good point. I added a pcall for glob matching (the "files" config) so if there are for example captures referenced that are not in the pattern this should not break neo-tree.
I don't know how to test for an invalid pattern (the "pattern" config) itself because I only managed to test an invalid string at all but then my whole nvim configuration breaks. If you have an example how I could test this case I will test it. But imho if it's an valid string then it's also a valid pattern.
- [x]  People that do not use file nesting will not be impacted at all

I also removed the can_have_nesting function and fixed a possible bug I found (variable is overridden locally which maybe could cause problems).

cseickel commented:
> * But imho if it's an valid string then it's also a valid pattern.

That's enough. If it syntactically incorrect then that's just on the user. I was more concerned with something like your first case around the captures where it could seem like a valid config but then cause a crash when used.

cseickel reviewed (approved):
All good, thanks!
